### Reduction operation: the sum of the numbers in the range [0, a)

In [1]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

Time taken by reduction operation: 0.03409147262573242 seconds
35.7 ms ± 2.95 ms per loop (mean ± std. dev. of 2 runs, 10 loops each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



In [4]:
print("="*60)
print("Ejercicio a): Utilizando listas de python")
print("="*60)

#creamos una lista con range
print("\n1. Creando lista con list(range(value))...")
start = time.time()
lista = list(range(value))
tiempo_creacion_lista = time.time() - start
print(f"   Tiempo de creación: {tiempo_creacion_lista:.6f} segundos") #le ponemos un format para ajustar n de decimales

#Método 1: Suma con bucle for
def reduc_operation_lista_for(a):
    """Suma elementos de una lista usando bucle for."""
    x = 0
    for num in a:
        x += num
    return x

print("\n2. Suma con bucle for sobre lista")
initialTime = time.time()
suma_lista_for = reduc_operation_lista_for(lista)
finalTime = time.time()
tiempo_lista_for = finalTime - initialTime
print(f"   Resultado: {suma_lista_for}")
print(f"   Tiempo: {tiempo_lista_for:.6f} segundos")

# Con %timeit
print("\n   Usando %timeit:")
%timeit -r 2 reduc_operation_lista_for(lista)

# Método 2: Suma con función sum()
def reduc_operation_lista_sum(a):
    """Suma elementos de una lista usando sum()."""
    return sum(a)

print("\n3. Suma con función sum() sobre lista")
initialTime = time.time()
suma_lista_sum = reduc_operation_lista_sum(lista)
finalTime = time.time()
tiempo_lista_sum = finalTime - initialTime
print(f"   Resultado: {suma_lista_sum}")
print(f"   Tiempo: {tiempo_lista_sum:.6f} segundos")

# Con timeit
print("\n   Usando %timeit:")
%timeit -r 2 reduc_operation_lista_sum(lista)


Ejercicio a): Utilizando listas de python

1. Creando lista con list(range(value))...
   Tiempo de creación: 0.022599 segundos

2. Suma con bucle for sobre lista
   Resultado: 499999500000
   Tiempo: 0.027474 segundos

   Usando %timeit:
26.3 ms ± 240 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

3. Suma con función sum() sobre lista
   Resultado: 499999500000
   Tiempo: 0.005943 segundos

   Usando %timeit:
5.73 ms ± 6.2 μs per loop (mean ± std. dev. of 2 runs, 100 loops each)


In [5]:
import numpy as np

print("\n" + "="*60)
print("Ejercicio B: Utilizando arrays de numpy")
print("="*60)

# Convertir lista a array de numpy
print("\n1. Convirtiendo lista a array de numpy...")
start = time.time()
array_numpy = np.array(lista)
tiempo_conversion = time.time() - start
print(f"   Tiempo de conversión: {tiempo_conversion:.6f} segundos")

# Método 1: Suma con bucle for sobre array de numpy
def reduc_operation_numpy_for(a):
    """Suma elementos de un array numpy usando bucle for."""
    x = 0
    for num in a:
        x += num
    return x

print("\n2. Suma con bucle for sobre array nupmy")
initialTime = time.time()
suma_numpy_for = reduc_operation_numpy_for(array_numpy)
finalTime = time.time()
tiempo_numpy_for = finalTime - initialTime
print(f"   Resultado: {suma_numpy_for}")
print(f"   Tiempo: {tiempo_numpy_for:.6f} segundos")

#Con %timeit
print("\n   Usando %timeit:")
%timeit -r 2 reduc_operation_numpy_for(array_numpy)

# Método 2: Suma con np.sum()
def reduc_operation_numpy_sum(a):
    """Suma elementos de un array NumPy usando np.sum()."""
    return np.sum(a)

print("\n3. Suma con np.sum() sobre array de numpy")
initialTime = time.time()
suma_numpy_sum = reduc_operation_numpy_sum(array_numpy)
finalTime = time.time()
tiempo_numpy_sum = finalTime - initialTime
print(f"   Resultado: {suma_numpy_sum}")
print(f"   Tiempo: {tiempo_numpy_sum:.6f} segundos")

# Con timeit
print("\n   Usando %timeit:")
%timeit -r 2 reduc_operation_numpy_sum(array_numpy)



Ejercicio B: Utilizando arrays de numpy

1. Convirtiendo lista a array de numpy...
   Tiempo de conversión: 0.045926 segundos

2. Suma con bucle for sobre array nupmy
   Resultado: 499999500000
   Tiempo: 0.051563 segundos

   Usando %timeit:
51.7 ms ± 116 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

3. Suma con np.sum() sobre array de numpy
   Resultado: 499999500000
   Tiempo: 0.000355 segundos

   Usando %timeit:
128 μs ± 361 ns per loop (mean ± std. dev. of 2 runs, 10,000 loops each)


## Análisis de resultados

### Código original vs Lista

El código original (range + for) tarda 0.034091 segundos, mientras que trabajar directamente sobre una lista ya creada tarda 0.022599 segundos. Esto ocurre porque:

- En el código original, "range(value)" se genera dinámicamente dentro del bucle.
- Con la lista, los datos ya están en memoria y listos para usar
- Sin embargo, crear la lista consume memoria adicional

**Conclusión**: La lista es más rápida en el bucle, pero ocupa mucha memoria.

### Mejora con función sum()

La función "sum()" reduce notablemente el tiempo a 0.005943 segundos. Esto se debe a:

- Implementación en C
- Optimizaciones internas del lenguaje
- Menor número de operaciones intermedias

**Conclusión**: Ejecutar "sum()" directamente es más rápido que meterlo en una función.

### Arrays numpy

**Con bucle for**: El tiempo aumenta respecto a listas
- Causa: Cada acceso a un elemento requiere conversión python a numy

**Con np.sum()**: El tiempo se reduce a la mitad respecto a listas

**Conclusión final**: numpy solo es eficiente cuando usas sus funciones vectorizadas. Los bucles for son contraproducentes.